src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/MetadataService.kt

```diff
@@ -0,0 +1,189 @@
+
+        runningLock = false
+    }
+
+    suspend fun indexMetadatas(requests: List<DocWriteRequest<*>>) {
+        if (requests.isEmpty()) return
+        var requestsToRetry = requests
+
+        logger.info("update config index mapping")
+        // when we try to index sth to config index
+        // we need to make sure the schema is up to date
+        if (!imIndices.attemptUpdateConfigIndexMapping()) {
+            logger.error("Failed to update config index mapping.")
+            return
+        }
+
+        retryPolicy.retry(logger, listOf(RestStatus.TOO_MANY_REQUESTS)) {
+            val bulkRequest = BulkRequest().add(requestsToRetry)
+            val bulkResponse: BulkResponse = client.suspendUntil { bulk(bulkRequest, it) }
+            val failedResponses = (bulkResponse.items ?: arrayOf()).filter { it.isFailed }
+
+            val retryIndexUuids = mutableListOf<Int>()
+            bulkResponse.items.forEach {
+                val indexUuid = revertISMMetadataID(it.id)
+                if (it.isFailed) {
+                    if (it.status() == RestStatus.TOO_MANY_REQUESTS) {
+                        retryIndexUuids.add(it.itemId)
+                    } else {
+                        logger.error("failed reason: ${it.failure}, ${it.failureMessage}")
+                        failedToIndexIndices[indexUuid] = it.failure
+                    }
+                } else {
+                    successIndexedIndices.add(indexUuid)
+                    failedToIndexIndices.remove(indexUuid)
+                }
+            }
+            requestsToRetry = retryIndexUuids.map { bulkRequest.requests()[it] }
+
+            if (requestsToRetry.isNotEmpty()) {
+                val retryCause = failedResponses.first { it.status() == RestStatus.TOO_MANY_REQUESTS }.failure.cause
+                throw ExceptionsHelper.convertToElastic(retryCause)
+            }
+        }
+    }
+
+    private suspend fun cleanMetadatas(indices: List<Index>) {
+        if (indices.isEmpty()) return
+
+        val request = UpdateManagedIndexMetaDataRequest(indicesToRemoveManagedIndexMetaDataFrom = indices)
+        try {
+            retryPolicy.retry(logger) {
+                val response: AcknowledgedResponse =
+                    client.suspendUntil { execute(UpdateManagedIndexMetaDataAction.INSTANCE, request, it) }
+                if (response.isAcknowledged) {
+                    failedToCleanIndices.removeAll(indices)
+                } else {
+                    logger.error("Failed to clean cluster state metadata for indices: [$indices].")
+                    failedToCleanIndices.addAll(indices)
+                }
+            }
+        } catch (e: ClusterBlockException) {
+            logger.error(
+                "Failed to clean cluster state metadata for indices: [$indices], due to cluster block exception.",
+                e
+            )
+            failedToCleanIndices.addAll(indices)
```

Comment:
If you have 1000 failing repeatedly is failedToCleanIndices a list that just keeps growing forever?

Reply:
Consider making it a set?

Reply:
changed to set